### PR TITLE
Avoid ending up with negative last page in REST Data Panache

### LINKED ISSUE #20424: "RestDataPanacheException: Failed to list the entities" when table is empty
**Describe the bug**
Using `@RepositoryRestResource` throws an exception when the table is empty.

**Expected behavior**
Call to `/tasks` should return an empty list of tasks, similar to this response with an empty list of countries.

image

**Actual behavior**
Call to `/tasks` triggers a `RestDataPanacheException` caused by: `java.lang.IllegalArgumentException: Page index must be >= 0 : -1`.

```
2021-09-27 11:54:18,683 WARN  [io.qua.spr.dat.res.run.RestDataPanacheExceptionMapper] (executor-thread-0) Mapping an unhandled RestDataPanacheException: io.quarkus.rest.data.panache.RestDataPanacheException: Failed to list the entities
	at org.acme.repository.TaskRepositoryJaxRs_47e783e09fc1ec6dc860f6e2c7cfb85a720cc58d.list(TaskRepositoryJaxRs_47e783e09fc1ec6dc860f6e2c7cfb85a720cc58d.zig:671)
	at org.acme.repository.TaskRepositoryJaxRs_47e783e09fc1ec6dc860f6e2c7cfb85a720cc58d_Subclass.list$$superforward1(TaskRepositoryJaxRs_47e783e09fc1ec6dc860f6e2c7cfb85a720cc58d_Subclass.zig:514)
	at org.acme.repository.TaskRepositoryJaxRs_47e783e09fc1ec6dc860f6e2c7cfb85a720cc58d_Subclass$$function$$6.apply(TaskRepositoryJaxRs_47e783e09fc1ec6dc860f6e2c7cfb85a720cc58d_Subclass$$function$$6.zig:57)
	at io.quarkus.arc.impl.AroundInvokeInvocationContext.proceed(AroundInvokeInvocationContext.java:54)
	at io.quarkus.arc.runtime.devconsole.InvocationInterceptor.proceed(InvocationInterceptor.java:62)
	at io.quarkus.arc.runtime.devconsole.InvocationInterceptor.monitor(InvocationInterceptor.java:49)
	at io.quarkus.arc.runtime.devconsole.InvocationInterceptor_Bean.intercept(InvocationInterceptor_Bean.zig:521)
	at io.quarkus.arc.impl.InterceptorInvocation.invoke(InterceptorInvocation.java:41)
	at io.quarkus.arc.impl.AroundInvokeInvocationContext.perform(AroundInvokeInvocationContext.java:41)
	at io.quarkus.arc.impl.InvocationContexts.performAroundInvoke(InvocationContexts.java:32)
	at org.acme.repository.TaskRepositoryJaxRs_47e783e09fc1ec6dc860f6e2c7cfb85a720cc58d_Subclass.list(TaskRepositoryJaxRs_47e783e09fc1ec6dc860f6e2c7cfb85a720cc58d_Subclass.zig:1115)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.jboss.resteasy.core.MethodInjectorImpl.invoke(MethodInjectorImpl.java:170)
	at org.jboss.resteasy.core.MethodInjectorImpl.invoke(MethodInjectorImpl.java:130)
	at org.jboss.resteasy.core.ResourceMethodInvoker.internalInvokeOnTarget(ResourceMethodInvoker.java:660)
	at org.jboss.resteasy.core.ResourceMethodInvoker.invokeOnTargetAfterFilter(ResourceMethodInvoker.java:524)
	at org.jboss.resteasy.core.ResourceMethodInvoker.lambda$invokeOnTarget$2(ResourceMethodInvoker.java:474)
	at org.jboss.resteasy.core.interception.jaxrs.PreMatchContainerRequestContext.filter(PreMatchContainerRequestContext.java:364)
	at org.jboss.resteasy.core.ResourceMethodInvoker.invokeOnTarget(ResourceMethodInvoker.java:476)
	at org.jboss.resteasy.core.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:434)
	at org.jboss.resteasy.core.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:408)
	at org.jboss.resteasy.core.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:69)
	at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:492)
	at org.jboss.resteasy.core.SynchronousDispatcher.lambda$invoke$4(SynchronousDispatcher.java:261)
	at org.jboss.resteasy.core.SynchronousDispatcher.lambda$preprocess$0(SynchronousDispatcher.java:161)
	at org.jboss.resteasy.core.interception.jaxrs.PreMatchContainerRequestContext.filter(PreMatchContainerRequestContext.java:364)
	at org.jboss.resteasy.core.SynchronousDispatcher.preprocess(SynchronousDispatcher.java:164)
	at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:247)
	at io.quarkus.resteasy.runtime.standalone.RequestDispatcher.service(RequestDispatcher.java:73)
	at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler.dispatch(VertxRequestHandler.java:138)
	at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler$1.run(VertxRequestHandler.java:93)
	at io.quarkus.vertx.core.runtime.VertxCoreRecorder$13.runWith(VertxCoreRecorder.java:543)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2449)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1478)
	at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:29)
	at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:29)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: java.lang.IllegalArgumentException: Page index must be >= 0 : -1
	at io.quarkus.panache.common.Page.<init>(Page.java:53)
	at io.quarkus.panache.common.Page.index(Page.java:118)
	at org.acme.repository.TaskRepositoryJaxRs_47e783e09fc1ec6dc860f6e2c7cfb85a720cc58d.list(TaskRepositoryJaxRs_47e783e09fc1ec6dc860f6e2c7cfb85a720cc58d.zig:376)
```

**To Reproduce**
1.  `git clone https://github.com/ethanimproving/empty-list-defect.git`
2. Create a MySQL Database called "scratch" and start the application. Ensure there is NO data in the task table.
3. Send a GET request to `http://localhost:8080/tasks`

**Environment (please complete the following information):**
 - Browser: Google Chrome
 - OS: Microsoft Windows 10 Enterprise

**Additional context**

The exception is handled in `quarkus-spring-data-rest`, so I suspect the issue is with that package.

image

Somehow the `list()` method of the JPA implementation of my TaskRepository (`TaskRepositoryJaxRs_47e783e09fc1ec6dc860f6e2c7cfb85a720cc58d`) is calling the `io.quarkus.panache.common` Pagination with a negative index. When my table is empty, my page index is -1 instead of 0.

image

### PR DESCRIPTION
**Pull request by gsmet**
Fixes #20424

### CODE REVIEW

**Comment by gastaldi** [comment]
If I understood this correctly, you're replacing:
```java
Integer.sum(pageCount, -1)
```

with

```java
Math.max(Integer.sum(pageCount,-1), 0)
```

**Review by gastaldi — APPROVED** [review]

File: extensions/panache/rest-data-panache/deployment/src/main/java/io/quarkus/rest/data/panache/deployment/utils/PaginationImplementor.java [hunk 1]
```diff
@@ -119,7 +119,12 @@ private ResultHandle getPageUri(BytecodeCreator creator, ResultHandle uriInfo, R
      */
     private ResultHandle getLastPage(BytecodeCreator creator, ResultHandle page, ResultHandle pageCount) {
         ResultHandle pageNumber = creator.invokeStaticMethod(
-                ofMethod(Integer.class, "sum", int.class, int.class, int.class), pageCount, creator.load(-1));
+                ofMethod(Math.class, "max", int.class, int.class, int.class),
+                creator.invokeStaticMethod(
+                        ofMethod(Integer.class, "sum", int.class, int.class, int.class),
+                        pageCount,
+                        creator.load(-1)),
```
Comment (gastaldi) [inline extensions/panache/rest-data-panache/deployment/src/main/java/io/quarkus/rest/data/panache/deployment/utils/PaginationImplementor.java:126]: Can we replace this snippet with the code below?

```suggestion
                creator.add(pageCount, creator.load(-1)),
```

Comment (gsmet) [reply extensions/panache/rest-data-panache/deployment/src/main/java/io/quarkus/rest/data/panache/deployment/utils/PaginationImplementor.java:126]: Maybe. But frankly, I prefer avoid touching it more than necessary.

**Comment by gsmet** [comment]
Yes, pages are 0 based. If you don't have any pages, the last page should still be 0.

**Pull request merged**